### PR TITLE
Open registration for c10 thread pool

### PR DESCRIPTION
**Pull request by jamesr66a**
1. Move ATen threadpool & open registration mechanism to C10
2. Move the `global_work_queue` to use this open registration mechanism, to allow users to substitute in their own

### CODE REVIEW

**Review by zdevito — APPROVED** [review]
OSS changes look fine.

File: c10/core/thread_pool.h [hunk 1]
```diff
@@ -100,8 +102,29 @@ class CAFFE2_API ThreadPool : public c10::TaskThreadPoolBase {
...
+C10_API TaskThreadPoolBase& global_work_queue();
+
+class C10_API TaskThreadPool : public c10::ThreadPool {
+ public:
+  explicit TaskThreadPool(
+      std::size_t pool_size,
+      int numa_node_id = -1)
+      : ThreadPool(pool_size, numa_node_id) {}
+
+  // TODO move this to ATen/core/thread_pool.h
```
Comment (dzhulgakov) [inline c10/core/thread_pool.h]: why?

File: c10/core/thread_pool.h [hunk 2]
```diff
@@ -100,8 +102,29 @@ class CAFFE2_API ThreadPool : public c10::TaskThreadPoolBase {
...
+  void init_thread() override {
```
Comment (dzhulgakov) [inline c10/core/thread_pool.h]: probably better move to .cpp file (and some includes too)

File: c10/core/thread_pool.cpp [hunk 3]
```diff
@@ -125,9 +124,32 @@ void setNumThreads(size_t v) {
...
+    int,
+    int,
+    bool);
+
+namespace {
+
+std::shared_ptr<TaskThreadPoolBase> createC10ThreadPool(
+    int device_id,
+    int pool_size,
+    bool create_new) {
```
Comment (dzhulgakov) [inline c10/core/thread_pool.cpp]: enforce that create_new is false (as that's the only one supported)

File: c10/core/thread_pool.h (hunk 1 above)
Comment (jamesr66a) [reply c10/core/thread_pool.h]: I thought I deleted this comment

**Pull request closed without merging**